src/connector/src/parser/mod.rs

```diff
@@ -487,7 +499,11 @@ impl<P: ByteStreamSourceParser> P {
             actor_id = source_info.actor_id,
             source_id = source_info.source_id.table_id()
         );
-
+        let data_stream = if let Some(rate_limit) = &self.source_ctx().source_ctrl_opts.rate_limit {
+            Box::pin(ensure_largest_at_rate_limit(data_stream, *rate_limit))
```

Comment:
Do all source connectors go through this code path? This looks nice and clear.

Reply:
> Do all source connectors go through this code path? This looks nice and clear.

Most of them, except `encode native` ones, which is from datagen.

Reply:
`include as` can carry the offset column into stream chunk, allowing the partition anywhere from the chunk.

Reply:
Hmm but doesn't `include as` depend on user specifying the offset column? By default it won't be included.
Then user can't rate limit at all.

Reply:
An alternative is to always include offset columns, but mark them as hidden.

Reply:
> An alternative is to always include offset columns, but mark them as hidden. 

Yeah, we can automatically add this hidden column for source or table with connector and prune it afterwards.